imports/izar_protocol_v3.leo

```diff
@@ -0,0 +1,270 @@
+        }
+
+        if keepers.k14 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        if keepers.k15 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        if keepers.k16 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        if keepers.k17 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        if keepers.k18 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        if keepers.k19 != unvalid_contract() {
+            accpeted += 1u32;
+        }
+
+        Mapping::set(izar_keeper, true, keepers);
+        Mapping::set(izar_config, 1u8, accpeted);
+    }
+
+    transition verify (from_chain_id: u32, signatures: Signatures20, keepers: Keepers20, payload: IzarRecvMsg) {
+        let (result, accepted): (scalar, u32) = verify_sigs(keepers, signatures, payload);
+        let nonce: Nonce = Nonce {
+            chain_id: from_chain_id,
+            nonce: payload.nonce,
+        };
+
+        return then finalize (nonce, result, accepted);
+    }
+
+    finalize verify (nonce: Nonce, keeper_digest: scalar, accepted: u32) {
+        // check nonce
+        assert(!Mapping::get_or_use(nonces, nonce, false));
+
+        // check keeper digest & accepted
+        assert(keeper_digest == Mapping::get(izar_digest, true));
+        let valid_num: u32 = Mapping::get(izar_config, 1u8);
+        assert(accepted >= valid_num / 3u32 * 2u32); // TODO: check this
+    }
+
+    struct Keepers20 {
+        k0: address,
+        k1: address,
+        k2: address,
+        k3: address,
+        k4: address,
+        k5: address,
+        k6: address,
+        k7: address,
+        k8: address,
+        k9: address,
+        k10: address,
+        k11: address,
+        k12: address,
+        k13: address,
+        k14: address,
+        k15: address,
+        k16: address,
+        k17: address,
+        k18: address,
+        k19: address,
+    }
+
+    struct Signatures20 {
+        s0: signature,
+        s1: signature,
+        s2: signature,
+        s3: signature,
+        s4: signature,
+        s5: signature,
+        s6: signature,
+        s7: signature,
+        s8: signature,
+        s9: signature,
+        s10: signature,
+        s11: signature,
+        s12: signature,
+        s13: signature,
+        s14: signature,
+        s15: signature,
+        s16: signature,
+        s17: signature,
+        s18: signature,
+        s19: signature,
+    }
+
+    struct IzarRecvMsg {
+        nonce: u128,
+        to_addr: address,
+        to_asset_addr: address,
+        amount: u128,
+    }
+
+    struct IzarCrossMsg {
+        to_chain_id: u32,
+        to_addr: field,
+        to_asset_addr: field,
+        amount: u128,
+        fee: u128,
+    }
+
+    function verify_sigs (keepers: Keepers20, signatures: Signatures20, msg: IzarRecvMsg) -> (scalar, u32) {
+        let accepted: u32 = 0u32;
+
+        if signatures.s0.verify(keepers.k0, msg) {
```

Comment:
No need to call `si.verify()` if `keepers.ki` is `aleo1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq3ljyzc` in the first place.

Reply:
I am not sure whether it will reduce the time taken for proving or not. It seems that whether a statement is executed or not, it will be proven. Right?

Reply:
It will reduce proving time, and aligns with the logic of `update_keeper`.

Reply:
Got it.